In [141]:
import pandas as pd
import networkx as nx
import math
import matplotlib.pyplot as plt
import networkx.algorithms.bipartite as bipartite
%matplotlib inline

1.Identify a large 2-node network dataset—you can start with a dataset in a repository.  Your data should meet the criteria that it consists of ties between and not within two (or more) distinct groups.

I'm using the following 2-node network dataset data set

Nashville Meetup Network
https://www.kaggle.com/stkbailey/nashville-meetup#member-to-group-edges.csv

In [142]:
# load the data
df = pd.read_csv("member-to-group-edges.csv")
df.head()

member_id  group_id  weight
0       2069  19277993       3
1     625050  19277993       2
2    1939496  19277993       1
3    2606806  19277993       4
4    3438546  19277993       1

In [143]:
member.dtypes

source    int64
target    int64
weight    int64
dtype: object

In [144]:
member = df.rename(columns={"member_id":"source","group_id":"target"})
member.head()

source    target  weight
0     2069  19277993       3
1   625050  19277993       2
2  1939496  19277993       1
3  2606806  19277993       4
4  3438546  19277993       1

In [145]:
# make two graph from this
member_graph = nx.from_pandas_edgelist(member, edge_attr=True)


2.Reduce the size of the network using a method such as the island method described in chapter 4 of social network analysis.

In [146]:
def trim_edges(g, weight=1):
    g2=nx.Graph()
    for f, to, edata in g.edges(data=True):
        if edata['weight'] > weight:
            g2.add_edge(f,to,weight=edata['weight'])
    return g2

def island_method(g, iterations=5):
    weights= [edata['weight'] for f,to,edata in g.edges(data=True)]
    mn=int(min(weights))
    mx=int(max(weights))
    #compute the size of the step, so we get a reasonable step in iterations
    step=int((mx-mn)/iterations)
    return [[threshold, trim_edges(g, threshold)] for threshold in range(mn,mx,step)]


In [151]:
islands=island_method(member_graph)

In [152]:
for i in islands:
    # print the threshold level, size of the graph, and number of connected components
    print(i[0], len(i[1]), len(list(nx.connected_component_subgraphs(i[1]))))

1 12017 58
40 311 88
79 127 50
118 46 22
157 17 8
196 8 4


3.What can you infer about each of the distinct groups?

A threshhold value of 1 results in a network of 12017 persons and 58 connected components.



A threshhold value of 40 results in a network of 311 persons and 88 connected components.


A threshhold value of 79 results in a network of 127 persons and 50 connected components.


A threshhold value of 118 results in a network of 46 persons and 22 connected components.

A threshhold value of 157 results in a network of 17 persons and 8 connected components.


A threshhold value of 196 results in a network of 8 persons and 4 connected components.


A threshhold value of 1 results in a network of 12017 persons and 58 connected components.
Single users meetup in group might be accidental.
Repeated meetups are more likely happedn between groups of people who have some common interest mosts.
Thresholding value 40 revals that only 311 nodes left in the islands. In this case, it's the highest meaningful threshold - the remaining 311 nodes are the people most actively involved in "Principles of Network Analysis with NetworkX" activity most.
With threshold 196 , size of graph 8 and connected component of 4 means the pople might be the people inovlved in the development of networkx.
